src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContentOfT.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if !NETCOREAPP
+using System.Diagnostics;
+#endif
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    internal sealed partial class JsonContent<TValue> : JsonContent
```

Comment:
I added this class in place of the following approved API in the linked issue (https://github.com/dotnet/runtime/issues/51544):

```diff
public sealed partial class JsonContent : HttpContent
{
+    public static JsonContent Create(object? inputValue, Type inputType, JsonSerializerContext context, MediaTypeHeaderValue mediaType = null) { throw null; }
+    public static JsonContent Create<T>(T inputValue, JsonTypeInfo<T> jsonValueInfo, MediaTypeHeaderValue mediaType = null) { throw null; }
}
```

Having a separate class allows us to override `HttpContent.SerializeToStreamAsync(Stream stream, TransportContext? context)` in a way that doesn't call options-based overloads of the serializer, which root all built-in converters and reflection logic (we don't want this for Blazor scenarios). I'm assuming that when ILLinker sees this derived type, it won't root the implementation on the base `JsonContext` class since it is not referenced.

If my assumption checks out, then the new proposal would be the following:

```diff
+ public sealed partial class JsonContent<TValue> : JsonContent
+ {
+    public static JsonContent<object?> Create(object? inputValue, Type inputType, JsonSerializerContext context, MediaTypeHeaderValue mediaType = null) { throw null; }
+    public static JsonContent<TValue> Create<T>(T inputValue, JsonTypeInfo<T> jsonValueInfo, MediaTypeHeaderValue mediaType = null) { throw null; }
+ }

- public sealed partial class JsonContent : System.Net.Http.HttpContent
+ public partial class JsonContent : System.Net.Http.HttpContent
{
    internal JsonContent() { }
    ...
}
```

Alternatively, we could make `JsonContent<TValue>` derive directly from `HttpContent`.

cc @eerhardt on linker behavior.

Reply:
@eerhardt is correct - statically we don't know which of the two overloads will be used. Linker figures out whether to keep virtual methods by:
* Is the type which implements the method instantiated (do we see a "new" for it somewhere)
* Is the virtual method called anywhere (not the specific implementation but any of the methods on the virtual override chain)

In this case both would be true and so linker would keep it. The method is obviously called and the base class is instantiated (since the derived class is). Currently linker is not smart enough to figure out that if only the derived class is instantiated, the base class's version of the method will never be used (unless it's explicitely called via `base.Method()`). I don't know if this is actually possible - never thought through all the implications of such behavior.

Reply:
<details>
<summary>
Thanks. Tested locally and verified that the linker does keep the implementation on the base class (click to view)
</summary>

```cs
using System.Text.Json;

namespace LinkerVirtualMethods
{
    class Program
    {
        static void Main(string[] args)
        {
            DerivedContent<int> content = new();
            content.DoWorkPublic();
        }
    }

    public abstract class Content
    {
        public virtual void DoWorkPublic() => DoWork();
        protected abstract void DoWork();
    }

    public class DerivedContent : Content
    {
        protected override void DoWork() => JsonSerializer.Serialize("Hello, World!");
    }

    public sealed class DerivedContent<T> : DerivedContent
    {
        protected override void DoWork() { }
    }
}
```
</details>

> (unless it's explicitely called via base.Method()). I don't know if this is actually possible - never thought through all the implications of such behavior.

This was my assumption. Is there any motivation or tracking issue to pursue this?

For the purpose of this PR and the source-gen feature, I'll make `JsonContent<T>` derive directly from `HttpContent`.